### PR TITLE
Allow signs to update price from worth

### LINKED ISSUE #2638: buy / sell signs with prices from worth.yml (not player-editable prices)
### Feature request

**Buy / Sell - Signs that respect the worth.yml**

For a working economy, buy and sell signs seem totally useless right now.
Why are they designed that players can write their own prices to it, when there is
a worth.yml available?

I know that worth.yml is for the /sell command - but why not for signs?
The trade sign is very useful. But I would like that my players can use buy and sell signs
that read and fill prices from the worth.yml. (and don't let players set prices)

**How the feature is useful**

Scenario:
- Minecraft server with economy
- starting balance = 0
- don't want the players to have to use commands (like /sell)
- players can make their own sell sign
- easy selling with right click on sign / any item in hand / stack or single (with prices from worth.yml)
- op's can make buy signs / define item / stack or single / prices taken from worth.yml
- trade signs are fine =)

Thank you for reading. And again: sell / buy signs are useless right now when you want to run an economy based on the worth.yml

Best regards
HAL

### PR DESCRIPTION
**Pull request by moxventura**
closes #2638

### CODE REVIEW

File: Essentials/src/com/earth2me/essentials/Settings.java [hunk 1]
```diff
@@ -713,6 +716,41 @@ public boolean isEcoDisabled() {
...
+    private boolean sellSignUpdatedToWorth = false;
+
+    private boolean _getSellSignUpdatedToWorth() {
+      return config.getBoolean("economy-update-sell-sign-to-worth", false);
+    }
+
+    @Override
+    public boolean isSellSignUpdatedToWorth() {
+      return sellSignUpdatedToWorth;
+    }
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/Settings.java]: These config options do not require a helper method, you can inline the config.getBoolean call
```suggestion
    @Override
    public boolean isBuySignUpdatedToWorth() {
      return config.getBoolean("economy-update-buy-sign-to-worth", false);
    }

    @Override
    public boolean isSellSignUpdatedToWorth() {
      return config.getBoolean("economy-update-sell-sign-to-worth", false);
    }
```

File: Essentials/src/com/earth2me/essentials/Settings.java [hunk 2]
```diff
@@ -547,6 +547,9 @@ public void reloadConfig() {
         economyLog = _isEcoLogEnabled();
         economyLogUpdate = _isEcoLogUpdateEnabled();
         economyDisabled = _isEcoDisabled();
+        sellSignUpdatedToWorth = _getSellSignUpdatedToWorth();
+        buySignUpdatedToWorth = _getBuySignUpdatedToWorth();
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/Settings.java]: Reflects changes below
```suggestion

```

File: Essentials/src/com/earth2me/essentials/Settings.java [hunk 3]
```diff
@@ -713,6 +716,41 @@ public boolean isEcoDisabled() {
...
+    private boolean _getSellSignUpdatedToWorth() {
+      return config.getBoolean("economy-update-sell-sign-to-worth", false);
+    }
+
+    @Override
+    public boolean isSellSignUpdatedToWorth() {
+      return sellSignUpdatedToWorth;
+    }
+
+    private BigDecimal economyBuyMultiplier = new BigDecimal("1.0");
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/Settings.java]: Don't initialize variables here
```suggestion
    private BigDecimal economyBuyMultiplier;
```

File: Essentials/src/messages_en.properties [hunk 4]
```diff
@@ -421,6 +421,7 @@ powerToolRemove=§6Command §c{0}§6 removed from §c{1}§6.
 powerToolRemoveAll=§6All commands removed from §c{0}§6.
 powerToolsDisabled=§6All of your power tools have been disabled.
 powerToolsEnabled=§6All of your power tools have been enabled.
+priceChanged=§6Price changed from {0} to {1}.
```
Comment (JRoy) [inline Essentials/src/messages_en.properties]: Please do not add this here, Crowdin will automatically do this
```suggestion
```

File: Essentials/src/messages_nl.properties [hunk 5]
```diff
@@ -421,6 +421,7 @@ powerToolRemove=§6Commando §c{0}§6 verwijderd van §c{1}§6.
 powerToolRemoveAll=§6Alle commandos zijn verwijderd van §c{0}§6.
 powerToolsDisabled=§6Al uw powertools zijn uitgeschakeld.
 powerToolsEnabled=§6Al uw powertools zijn ingeschakeld.
+priceChanged=§6Prijs is veranderd van {0} naar {1}.
```
Comment (JRoy) [inline Essentials/src/messages_nl.properties]: Please do not add this here, Crowdin will automatically do this
```suggestion
```

File: Essentials/src/messages.properties [hunk 6]
```diff
@@ -427,6 +427,7 @@ powerToolRemove=\u00a76Command \u00a7c{0}\u00a76 removed from \u00a7c{1}\u00a76.
 powerToolRemoveAll=\u00a76All commands removed from \u00a7c{0}\u00a76.
 powerToolsDisabled=\u00a76All of your power tools have been disabled.
 powerToolsEnabled=\u00a76All of your power tools have been enabled.
+priceChanged=\u00a76Price changed from {0} to {1}.
```
Comment (JRoy) [inline Essentials/src/messages.properties]: I feel like this message could be a little more useful/explanatory

File: Essentials/src/config.yml [hunk 7]
```diff
@@ -692,6 +692,16 @@ max-money: 10000000000000
...
 min-money: -10000
 
+# Enable this to make sell signs automatically adjust their price according to worth.yml
+economy-update-sell-sign-to-worth: false
+
+# Enable this to make buy signs automatically adjust their price according to worth.yml
+economy-update-buy-sign-to-worth: false
+
+# If you don't want buy and sell prices to be the same, set this multiplier to something higher then 1.0
+# All buy prices for players will be multiplied by this amount.
```
Comment (JRoy) [inline Essentials/src/config.yml]: This description should be improved. It's not clear if this only happens when the above booleans are enabled.

File: Essentials/src/config.yml [hunk 8]
```diff
@@ -692,6 +692,16 @@ max-money: 10000000000000
 # Setting this to 0, will disable overdrafts/loans completely.  Users need 'essentials.eco.loan' perm to go below 0.
 min-money: -10000
 
+# Enable this to make sell signs automatically adjust their price according to worth.yml
+economy-update-sell-sign-to-worth: false
+
+# Enable this to make buy signs automatically adjust their price according to worth.yml
+economy-update-buy-sign-to-worth: false
```
Comment (JRoy) [inline Essentials/src/config.yml]: Not a fan of the descriptions here, 'automatically' implies that all signs would update upon reloading

File: Essentials/src/com/earth2me/essentials/signs/SignSell.java [hunk 9]
```diff
@@ -18,13 +18,24 @@ public SignSell() {
...
     protected boolean onSignCreate(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException {
-        validateTrade(sign, 1, 2, player, ess);
+        validateTrade(sign, 1, 2, player, ess);  
+        if(ess.getSettings().isSellSignUpdatedToWorth()) {
+            try {
+                updateFromWorth(sign, ess);
+            }
+            catch (SignException e) {
+                // It's fine
+            }
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/signs/SignSell.java]: No, it's not fine?

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java [hunk 10]
```diff
@@ -18,12 +17,23 @@ public SignBuy() {
...
     protected boolean onSignCreate(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException {
         validateTrade(sign, 1, 2, player, ess);
+        if(ess.getSettings().isBuySignUpdatedToWorth()) {
+            try {
+              updateFromWorth(sign, ess, ess.getSettings().getEcoBuyMultiplier());
+            }
+            catch (SignException e) {
+              // It's fine
+            }
+        }
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: No, it's not fine?

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 11]
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
...
+      BigDecimal price = ess.getWorth().getPrice(ess, stack);
+      if (price == null || amount == 0) return;
+      price = price.multiply(multiplier).multiply(new BigDecimal(amount));
+      final BigDecimal oldPrice = getMoney(getSignText(sign, 3), ess);
+      if ( oldPrice == null || price.compareTo(oldPrice) != 0) {
+        sign.setLine(3, NumberUtil.shortCurrency(price, ess));
+        sign.updateSign();
+        throw new SignException(tl("priceChanged", oldPrice, price));
+      }
+  }
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: Why the two separate methods here?

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 12]
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
...
+    }
+    
+    /**
+     * Updates the sign cost to the value of the item specified in worth.yml multiplied by a multiplier factor
+     */
+    protected void updateFromWorth(final ISign sign, final IEssentials ess, final BigDecimal multiplier) throws SignException {
+      final ItemStack stack = getItemStack(getSignText(sign, 2), getIntegerPositive(sign.getLine(1)), ess);
+      final int amount = stack.getAmount();
+      BigDecimal price = ess.getWorth().getPrice(ess, stack);
+      if (price == null || amount == 0) return;
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]:
```suggestion
      if (price == null || amount == 0) {
        return;
      }
```

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 13]
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
...
+      price = price.multiply(multiplier).multiply(new BigDecimal(amount));
+      final BigDecimal oldPrice = getMoney(getSignText(sign, 3), ess);
+      if ( oldPrice == null || price.compareTo(oldPrice) != 0) {
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]:
```suggestion
      if (oldPrice == null || price.compareTo(oldPrice) != 0) {
```

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 14]
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
...
+        sign.setLine(3, NumberUtil.shortCurrency(price, ess));
+        sign.updateSign();
+        throw new SignException(tl("priceChanged", oldPrice, price));
```
Comment (JRoy) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: The message will never be sent if you always ignore it.

**Review by JRoy — CHANGES_REQUESTED** [review]
Overall not a fan of this feature but would be happy to hear the input of others @md678685 @pop4959

**Comment by moxventura** [comment]
Thanks for the review! I'll fix the issues..

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 10 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: In this location this means that the price could not be found in worth.yml If we do nothing here the sign will work as it used to. Should I elaborate on that in the comment?

File: Essentials/src/com/earth2me/essentials/Settings.java (hunk 1 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/Settings.java]: I was wondering where those helper functions were for :)

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java (hunk 14 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: if I test this I see this message. The exception is caught somewhere and shown to the player. I assumed this was how it is supposed to work. Maybe it is a more elegant solution if I catch this somewhere else and display the message there?

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java (hunk 11 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: A habit, I can change it to one...

File: Essentials/src/config.yml (hunk 8 above)
Comment (moxventura) [reply Essentials/src/config.yml]: I'll change this

File: Essentials/src/config.yml (hunk 7 above)
Comment (moxventura) [reply Essentials/src/config.yml]: true, I'll fix it. This multiplier could also be used if one would implement the 'buy' method (like in #2795). But for now I'll add more information to it.

File: Essentials/src/messages.properties (hunk 6 above)
Comment (moxventura) [reply Essentials/src/messages.properties]: I could add the entity name

File: Essentials/src/messages_en.properties (hunk 4 above)
Comment (moxventura) [reply Essentials/src/messages_en.properties]: didn't know that

**Review by pop4959 — CHANGES_REQUESTED** [review]
@JRoy I'm not so fond of this feature either, and frankly not sure why we haven't closed the original issue yet. There are several problems, both logistically with how the signs are meant to work, as well as with implementing it.

First of all, I want to mention that in the original issue I suggested that this would be a tricky feature to support since Essentials would need to keep track of buy/sell signs that are created. This would need to be done so that prices can be updated automatically. I do appreciate the fact that @moxventura attempted to go with a storage-less implementation, however it is very apparent that the signs 1) do not update automatically, and thus 2) will confuse or aggravate players trying to use them, since they will receive an "error" about the price changing right as they try to use it.

Now, to point out a few flaws with the feature request itself. It suggests a use case in which players set up their own buy/sell signs, which to me sounds like complete nonsense. These signs are intended to be created by server admins only, as they contain an infinite stock. Pulling automatically from worth isn't very useful, since usually prices are set higher than their normal sell price (which in this PR's implementation is re-solved by adding another setting for buy multiplier). Obviously the suggestion was to change the behavior so that players can create the signs without being able to set the price themselves (such as a zero amount) and bypassing worth. However, there is already the /sell command which meets this purpose (which the person suggesting the issue was deliberately trying to avoid using for whatever reason). Not to mention that requiring players to set up a sell sign just to sell an item, instead of using the /sell command just seems plain silly...

There is nothing wrong with trade signs, which are meant to be set up and used by players. They can set their own prices, but this is fine because the items they are selling are theirs to sell.

I'd be curious to hear more from @moxventura about why they were interested in making this PR. Personal interest, or just to help tackle some Essentials issues? (which, by the way is very much appreciated!) For future PRs, I'd advise checking in with a team member before making a PR for a larger feature (especially one that isn't tagged with "open to pr"), especially when there have been concerns voiced around it.

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java (hunk 14 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: This is caught on line 64 and 118, I think that's fine

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 10 above)
Comment (JRoy) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: There should be a check in the method to validate this, not a try catch here

**Comment by moxventura** [comment]
@pop4959 thanks for the reply! I'd love to elaborate. First of all, I agree completely with you that it's not the goal to have signs that are placed by the players itself. It was suggested in the original post but I agree that the signs should be placed by ops. 
I was triggered to create this PR due to the comment of @Technoguyfication in the issue since I had the same problem myself. 

The goal is to allow easy change of global prices without having moderators change all the signs in your world and find out you forgot one or two which people are abusing to gain infinite money. This will allow moderators to balance the economy during gameplay on a global scale depending on the needs of the server. (Brainfart: ultimately you could let the supply and demand calculate prices and create a real simulation of an economy. Maybe I'll get my hands on this idea later on).

My second goal was to leave the original operation in tact so this is an added feature which is turned off by default and does no harm to people who won't use it.

The way the current PR is implemented will allow server admins to do a single command (/setworth [item] [price]) and the prices will change on all signs. It's true that the first user who tries to use a sign with an old price will get a message of the failed transfer, but it will be clear the price has changed and after that first contact the price on the sign will be correct and the sign will function as expected. But to be honest, this is how an economy works.. prices change.

I'd love to help out with other issues as well, I had a lot of fun figuring out how to create this one and I already use it on my server. Also having my code reviewed is really educational. Thanks!

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 10 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: Could you help me out how I should do that? 
The solution in my head now is to pass a boolean whether or not an exception should be thrown. But I'm not fond if this idea. I'll have to know the difference between a 'create' update and a 'use' update if I want to be able to make the decision whether or not to throw the exception.

**Comment by Techno3600** [comment]
Hey, so while some might find this feature useless. I certainly do not. I don’t think many understand the pain of going through an entire worth.yml only to find you have to also hardcode those values into the signs. I haven’t tested this out but these should only work with admin shop signs.

Also, could you make an option to have these only affect the sell line in signs. Currently, my server marks up the buy prices of items so it would be perfect to just use this and then manually mark up the buy price.

Also, I understand the concern that this would be frustrating to players when it updates on the spot, but my management would just go through and hit all the signs in my shop once (because we know what has admin signs from the worth.yml).

I understand why some don’t like this idea. But I think there is a certain convenience factor with this change that it’s almost dumb not to implement it.

I am a programmer but haven’t delved deep into Minecraft plugins yet. Would it be too difficult to find all the signs in the world, check to see if they’re valid admin shop signs, and then update them to the correct prices? If that’s not possible, the current idea you have where it updates on click is fine for me since we’re going to hit it anyway before a change goes live.

Thanks,
Techno

**Comment by moxventura** [comment]
Hey, thanks for the feedback. If you are interested I could send you the compiled jar (or you could compile it yourself) to see how it works. I really think players won't mind seeing the price-change message.

> Also, could you make an option to have these only affect the sell line in signs. Currently, my server marks up the buy prices of items so it would be perfect to just use this and then manually mark up the buy price.

I'm not sure what you mean, but if you mean only sell-signs then the feature is already in there. These are the config settings I added:
```
# Enable this to make sell signs adjust their price according to worth.yml when they are created or used
economy-update-sell-sign-to-worth: false

# Enable this to make buy signs adjust their price according to worth.yml when they are created or used
economy-update-buy-sign-to-worth: false

# If you don't want buy and sell prices to be the same, set this multiplier to something higher then 1.0
# All buy prices on buy signs will be the worth value multiplied by this amount.
economy-buy-multiplier: 1.0
```
So you could only set this feature for the sell signs.
I did not touch the trade-signs...

> I am a programmer but haven’t delved deep into Minecraft plugins yet. Would it be too difficult to find all the signs in the world, check to see if they’re valid admin shop signs, and then update them to the correct prices? If that’s not possible, the current idea you have where it updates on click is fine for me since we’re going to hit it anyway before a change goes live.

I did play with this idea for a while but from what I understood is that it would mean to check every block in the server to see if it's a sign. And you'll need to do that every time a price has changed. That's a performance hit, even if you do it on chunk-load. You might be better off doing it offline using a level editor. Another option would be to keep a list of signs somewhere, but that's a lot of work I think.

**Comment by Technoguyfication** [comment]
Maybe we could differentiate signs that have a fixed worth and signs that have a dynamic worth from `/setworth` or `worth.yml` by putting the price on the sign in brackets, for example. This would let old signs function as normal and keep their old prices, and the plugin can also check to see if line 4 of the sign is inside brackets before updating the sign price. That way, there's no "all-or-nothing" option in the config to convert _all_ signs in the world, and instead let admins choose which signs update automatically.

Example:
```
[Buy]
64
cobblestone
$64
```
...would act as a normal sign and the price of $64 never changes. For a dynamic sign, it looks like this:
```
[Buy]
64
cobblestone
[$64]
```
This sign will update automatically. If the price of cobblestone is set to $2, the next time the sign is interacted with, it will update to `[$128]`

**Comment by pop4959** [comment]
I don't have time to give a full response currently but I'm glad we're on the same page about buy/sell signs clearly being an admin feature. Also, I really do think the implementation of this could be better. I was planning to suggest something similar to what @Technoguyfication just mentioned, with a sign syntax that indicates the price is set by worth. This would allow an admin to selectively create signs that use worth, while also not limiting them to only using worth.

**Comment by Techno3600** [comment]
I am so happy you also have a buy multiplayer. You have no idea how frustrating it was for our management team to have to make a whole worth.yml and then hardcode all the values in. I think they will be happy to hear a change is in the works!

**Comment by Techno3600** [comment]
@moxventura Also, when a user clicks on the sign when it's not updated to the new price in the worth.yml, will it charge them the new amount and then tell them the price has changed? Or will it cancel the transaction, tell them the price has changed, update the sign, and then allow them to recomplete the transaction?

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 10 above)
Comment (Technoguyfication) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: In general, it's bad practice to throw exceptions during normal operation of a program. They're computationally expensive and make the code harder to understand when an exception is "supposed" to be thrown. You should return a value like a boolean that indicates whether the price changed, and then if the price _did_ change, cancel the buy/sell and let the player know about the price change so they can decide whether to complete their transaction at the new price.

**Comment by moxventura** [comment]
> @moxventura Also, when a user clicks on the sign when it's not updated to the new price in the worth.yml, will it charge them the new amount and then tell them the price has changed? Or will it cancel the transaction, tell them the price has changed, update the sign, and then allow them to recomplete the transaction?

@Techno3600 The latter, let's give them the opportunity to reconsider their purchase. 
[](https://postimg.cc/9DLdc7Tf)

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 10 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: Fixed

**Comment by moxventura** [comment]
Ok, so here's what I've come up with. @Technoguyfication @pop4959 

If you leave the 'cost' line empty on placing a sign: it will automatically use the price from worth.yml
Otherwise it will use the cost which is provided.
If the item is not found in worth.yml it will keep the current behaviour (which is buy/sell the items for free, which might be a bug?)

Anyway, if a price from worth.yml is used, the price is colored (yellow). And will update the price as soon as the price has changed. Also see the attached gif:

https://i.postimg.cc/MK14NT6J/mc-signs.gif

File: Essentials/src/config.yml [hunk 15]
```diff
@@ -709,6 +709,10 @@ max-money: 10000000000000
 # Setting this to 0, will disable overdrafts/loans completely.  Users need 'essentials.eco.loan' perm to go below 0.
 min-money: -10000
 
+# If you don't want buy and sell prices to be the same, set this multiplier to something higher then 1.0
+# All buy prices on buy signs will be the worth value multiplied by this amount.
+economy-buy-multiplier: 1.0
```
Comment (pop4959) [inline Essentials/src/config.yml]: maybe just "buy-multiplier" since this is already in the economy section

File: Essentials/src/messages.properties [hunk 16]
```diff
@@ -583,6 +583,7 @@ powerToolRemove=\u00a76Command \u00a7c{0}\u00a76 removed from \u00a7c{1}\u00a76.
 powerToolRemoveAll=\u00a76All commands removed from \u00a7c{0}\u00a76.
 powerToolsDisabled=\u00a76All of your power tools have been disabled.
 powerToolsEnabled=\u00a76All of your power tools have been enabled.
+priceChanged=\u00a76The price of \u00a7c{0}x {1}\u00a76 has changed from \u00a7e{2}\u00a76 to \u00a7e{3}\u00a76.
```
Comment (pop4959) [inline Essentials/src/messages.properties]: should probably not be displayed as an error, since it isn't one. also, normally c is used for variables, not e.

**Review by pop4959 — CHANGES_REQUESTED** [review]
Not a full review, just some more comments

File: Essentials/src/messages.properties (hunk 16 above)
Comment (moxventura) [reply Essentials/src/messages.properties]: changed

File: Essentials/src/config.yml (hunk 15 above)
Comment (moxventura) [reply Essentials/src/config.yml]: makes sense

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 17]
```diff
@@ -375,7 +376,8 @@ protected final ItemStack getItemMeta(final ItemStack item, final String meta, f
         return stack;
     }
 
-    protected final BigDecimal getMoney(final String line, final IEssentials ess) throws SignException {
+    protected final BigDecimal getMoney(String line, final IEssentials ess) throws SignException {
+        line = ChatColor.stripColor(line);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:379]: Why is this change here?

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 18]
```diff
@@ -435,6 +437,33 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
...
+            if (oldPrice == null || price.compareTo(oldPrice) != 0) {
+                final String priceString = NumberUtil.shortCurrency(price, ess);
+                sign.setLine(3, tl("signFormatWorth") + priceString);
+                sign.updateSign();
+                player.sendMessage(tl("priceChanged", amount, stack.getType().toString().toLowerCase(Locale.ENGLISH), (oldPrice == null ? tl("none") : NumberUtil.shortCurrency(oldPrice, ess)), priceString));
+                return true;
+            }
+        }
+        return false;
+  }
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: This behaviour is specific to buy and sell signs - I'd be inclined to put this in an `EssentialsShopSign` subclass.

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java [hunk 19]
```diff
@@ -18,12 +19,20 @@ public SignBuy() {
     @Override
     protected boolean onSignCreate(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException {
         validateTrade(sign, 1, 2, player, ess);
-        validateTrade(sign, 3, ess);
+        if(sign.getLine(3).isEmpty()) {
+            updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier());
+        }
+        else {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignBuy.java]:
```suggestion
        if (sign.getLine(3).isEmpty()) {
            updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier());
        } else {
```

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java [hunk 20]
```diff
@@ -18,12 +19,20 @@ public SignBuy() {
...
+            validateTrade(sign, 3, ess);
+        }
         return true;
     }
 
     @Override
     protected boolean onSignInteract(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException, ChargeException, MaxMoneyException {
+        if(sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier())) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignBuy.java]:
```suggestion
        if (sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier())) {
```

File: Essentials/src/com/earth2me/essentials/signs/SignSell.java [hunk 21]
```diff
@@ -18,13 +20,21 @@ public SignSell() {
 
     @Override
     protected boolean onSignCreate(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException {
-        validateTrade(sign, 1, 2, player, ess);
-        validateTrade(sign, 3, ess);
+        validateTrade(sign, 1, 2, player, ess);  
+        if(sign.getLine(3).isEmpty()) {
+            updateFromWorth(sign, player, ess, new BigDecimal("1.0"));
+        }
+        else {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignSell.java]:
```suggestion
        if (sign.getLine(3).isEmpty()) {
            updateFromWorth(sign, player, ess, new BigDecimal("1.0"));
        } else {
```

This is also common functionality between buy and sell which ought to be in a separate class.

File: Essentials/src/com/earth2me/essentials/signs/SignSell.java [hunk 22]
```diff
@@ -18,13 +20,21 @@ public SignSell() {
...
+            validateTrade(sign, 3, ess);
+        }
         return true;
     }
 
     @Override
     protected boolean onSignInteract(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException, ChargeException, MaxMoneyException {
+        if(sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, new BigDecimal("1.0"))) {
+            return false;
+        }
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignSell.java]:
```suggestion
        if (sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, new BigDecimal("1.0"))) {
            return false;
        }
```

Ditto (also, use an overloaded method for the default scaling factor).

File: Essentials/src/messages.properties [hunk 23]
```diff
@@ -592,6 +593,7 @@ ptimeCommandDescription=Adjust player's client time. Add @ prefix to fix.
 ptimeCommandUsage=/<command> [list|reset|day|night|dawn|17:30|4pm|4000ticks] [player|*]
 pweatherCommandDescription=Adjust a player's weather
 pweatherCommandUsage=/<command> [list|reset|storm|sun|clear] [player|*]
+priceChanged=\u00a76The price of \u00a7c{0}x {1}\u00a76 has changed from \u00a7e{2}\u00a76 to \u00a7e{3}\u00a76.
```
Comment (mdcfe) [inline Essentials/src/messages.properties]: We should prompt the player to right-click the sign again if they want to buy/sell at the new price.

File: Essentials/src/messages.properties [hunk 24]
```diff
@@ -584,6 +584,7 @@ powerToolRemove=\u00a76Command \u00a7c{0}\u00a76 removed from \u00a7c{1}\u00a76.
 powerToolRemoveAll=\u00a76All commands removed from \u00a7c{0}\u00a76.
 powerToolsDisabled=\u00a76All of your power tools have been disabled.
 powerToolsEnabled=\u00a76All of your power tools have been enabled.
+priceChanged=\u00a76The price of \u00a7c{0}x {1}\u00a76 has changed from \u00a7c{2}\u00a76 to \u00a7c{3}\u00a76.
```
Comment (mdcfe) [inline Essentials/src/messages.properties]:
```suggestion
```

Duplicate message

**Review by mdcfe — CHANGES_REQUESTED** [review]

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 19 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: you cuddle the else? It's a matter of taste I guess :)

Comment (pop4959) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: you mean the standard format that everyone uses? lol

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java (hunk 17 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:379]: I color the lines if it's updated from worth. So you can see which signs use the values from worth.yml. If I don't strip the color here we won't be able to get the price value since there could be non-number (color) characters on this line.

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java (hunk 18 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: Done

File: Essentials/src/messages.properties (hunk 24 above)
Comment (moxventura) [reply Essentials/src/messages.properties]: Must have happened with a merge conflict 👎

File: Essentials/src/messages.properties (hunk 23 above)
Comment (moxventura) [reply Essentials/src/messages.properties]: Done

File: Essentials/src/com/earth2me/essentials/signs/SignBuy.java (hunk 19 above)
Comment (moxventura) [reply Essentials/src/com/earth2me/essentials/signs/SignBuy.java]: Guess I'm old school ¯\_(ツ)_/¯

**Comment by moxventura** [comment]
I'm willing to update this again, but only if the choice is made to implement this feature.

**Comment by JRoy** [comment]
Pretty sure the consensus on this PR is that in order to do this perfectly, Essentials would need to store the locations of all signs and that is something we don't want to do. This solution of updating the signs on click is also not a good solution as it would be misleading to users. Thanks for your interest in contributing but we will reject this PR at this time

**Comment by Techno3600** [comment]
With all due respect, I completely disagree with this decision. This option is editable in the configuration, and thus people who really need this for their server could enable it, knowing that it will not update immediately. The alternative was hours upon hours of work for my team manually updating signs every time we made a change in worth.yml back when we had that season running. Our team, and I'm sure a lot of others, would easily take the option of signs updating onclick over hours of manually editing sings. Also, the multiplayer addition in this PR would have made our team's life so much easier because we wouldn't have had to pull out a calculator literally every single time we wanted to update the price.

I would strongly urge @JRoy and the other creators of EssentialsX to reconsider this decision and possibly add this into the master project.
